Added tests and tweaks to code

mengqvist opened the pull request:
Basically all my changes in one big pull request. Sorry about that.

Commit:
add tests

Commit:
change asserts to raise ValueError, its more stable in production

Commit:
add pytest

Commit:
add docstrings and argument typing

Commit:
added tests using values from paper, these are all failing

Commit:
add docstrings and typing

Commit:
add tests to ensure that energies are calculated correctly

Commit:
add source for the energy tables

Commit:
refactor code, add back comment, working on isolating bug

Commit:
add better printouts for tests

Commit:
add anoter test with published data

Commit:
add acylation, amidation, tweak how energies are calculated

Commit:
change hbond to -6

Commit:
add even more tests

Commit:
re-arrange code

Commit:
change print to ensure that correct method is printed

Commit:
description of data files

Commit:
begin refactoring of code to make energy contrubutions more transparent

Commit:
refactor energy functions to take sequence as input and output per-residue energies

Commit:
Merge branch 'main' of github.com:mengqvist/AGADIR into bugfix

Commit:
remove multiplication with 100

Commit:
commit notebooks usede for testing

Commit:
remove pip install info

Commit:
fix argument typing, remove unneeded code

Commit:
remove notebooks

Commit:
remove developers note

Commit:
transiition to lacroix energy tables

Commit:
add residue naming convention

Commit:
add aa index to tables, improve description

Commit:
move energies to separate script

Commit:
move verification functions to separate script

Commit:
first implementation of intrinsic energy and capping

Commit:
add code for hydrophobic staple motif

Commit:
add schellman motif

Commit:
fix issue with spaces

Commit:
adapt i+3 and i+4 to new table

mengqvist committed:
Merge pull request #1 from mengqvist/lacroix

A draft implementation using table 1, 2, 3, and 4 from the Lacroix paper

Commit:
fix varable names

Commit:
Merge branch 'main' of github.com:mengqvist/AGADIR into lacroix

hussius committed:
Isnt it better to have ddG in the _calc_dG_Hel function so we get it in the printout?

mengqvist committed:
Merge pull request #2 from mengqvist/hotfix/move_ddG

Isnt it better to have ddG in the _calc_dG_Hel function so we get it …

hussius committed:
Update Hbond value, add simple test script

mengqvist committed:
Merge pull request #3 from mengqvist/hotfix/update_hbond

Update Hbond value, add simple test script

reisalex commented:
Hi @mengqvist I'll review the changes this weekend! Looks great!

hussius committed:
First-pass equations for interaction energy

hussius committed:
First-pass calculations for interaction energy

hussius committed:
Add pH to AGADIR object

rcrehuet committed:
Update debye_huckel_full

ionic strength in mol/m**3

rcrehuet committed:
Update calculate_r

N+1 if index start from 0

reisalex commented:
Hi @mengqvist 

Could you please review for me this pytest that I have added?
https://github.com/reisalex/pyAGADIR/blob/main/tests/test_model_NSB_1994.py

This is Table 3 from the original Munoz and Serrano 1994 paper. This list of peptides and their CD measurements are a good test for us as we explore trying to deploy newer versions of AGADIR. My initial implementation does a reasonable job at predicting these measurements (R^2=0.45) with predicted percent helixes quite on par with the measurements. 

I noticed your PR model changes have a tendency to _overpredict_ the percent helix. Could you please review?

Some example output:
```
---
1s
Peptide: AVGKSNLLSRYARNEFSA
Measured: 2.00, Predicted: 26.83
---
1s
Peptide: RFRAVTSAYYRGAVG
Measured: 3.00, Predicted: 34.57
---
1s
Peptide: TRRTTFESVGRWLDELKIHSD
Measured: 7.50, Predicted: 20.98
---
1s
Peptide: AVSVEEGKALAEEEGLF
Measured: 4.00, Predicted: 1.31
---
1s
Peptide: STNVKTAFEMVILDIYNNV
Measured: 3.00, Predicted: 36.22
---
1s
Peptide: DTYKLILNGKTLKGETTTEA
Measured: 2.00, Predicted: 23.30
```

rcrehuet committed:
Added reference for calculate_r

This gives the linear dependence of the distance, which is explained in the text but not in an equation.

hussius committed:
Merge pull request #6 from rcrehuet/patch-3

Added reference for calculate_r

hussius committed:
Merge pull request #5 from rcrehuet/patch-2

Update calculate_r

hussius committed:
Merge pull request #4 from rcrehuet/patch-1

Update debye_huckel_full

Commit:
add amino acid pka values

Commit:
modify ddG energies

Commit:
change imports, remove unneeded function

Commit:
fix equation 12 merge conflicts

Commit:
fix hbond merge conflict

Commit:
move helix energy print statement

mengqvist committed:
Merge pull request #7 from mengqvist/lacroix

Add table with (temporary) pKa values, fix merge conflicts

Commit:
remove simble DH function

Commit:
ammend print statement

Commit:
add argument typing, pass temperature argument, load pKa from table

Commit:
fix variable names, acid and base were confused

Commit:
do temperature conversion at beginning of model, pass temperature to function

mengqvist committed:
Merge pull request #8 from mengqvist/feature/interaction

Add functions for calculating energy between the charged termini and the helix macrodipole.

rcrehuet committed:
Corrected value for dielectric constant at 273K

rcrehuet committed:
Corrected calculate_interaction_energy.py

rcrehuet committed:
Corrected N and C charge and interaction with dipole

rcrehuet committed:
Feature/interaction (#9)

* Corrected value for dielectric constant at 273K

* Corrected calculate_interaction_energy.py

* Corrected N and C charge and interaction with dipole

Commit:
added questions / todo

Commit:
Add question about capping

mengqvist committed:
Added distance tables 6 and 7 (#11)

* add missing data tables

* documentation for new tables

* parsing distance tables

rcrehuet committed:
added temperature dependence of water dielectric constant

hussius committed:
Merge branch 'main' into feature/interaction

hussius committed:
Merge pull request #12 from rcrehuet/feature/interaction

Feature/interaction

hussius committed:
Electrostatics (#13)

* add missing data tables

* documentation for new tables

* parsing distance tables

* (WIP) add dG_electrost term

* Add another version of Debye-Hückel and new fcn for electrostatic interactions

* Bug fix

* Bug fix Å->m

* Energy conversion

* Updates to electrostatics

---------

Co-authored-by: Martin Engqvist <martin@enginzyme.com>

mengqvist committed:
Features/electrostatics (#14)

* add missing data tables

* documentation for new tables

* parsing distance tables

* (WIP) add dG_electrost term

* Add another version of Debye-Hückel and new fcn for electrostatic interactions

* Bug fix

* Bug fix Å->m

* Energy conversion

* Updates to electrostatics

* change a few of the docstrings

* rename functions and variables for (hopefully) increased clarity

---------

Co-authored-by: Mikael Huss <mikael.huss@gmail.com>

mengqvist committed:
Adding missing pH-dependent terms and side-chain dipole interaction (#15)

* re-arrange calculate_term_dipole_interaction_energy for even more clarity

* add function to calculate ionic strength

* add pka for R

* add function to replicate figure 3b from lacroix 1998

* update todo

* add section on capping

* add function for charged side-chain macrodipole interaction

* add i3 and i4 energy scaling based on charged state

* add pH-dependent selection between 'Ncen' and 'Neutral'

* pass pH and temp where needed

* ensure that the macrodipole-terminal interaction only happens in uncapped peptides

* remove i+1 as these are likely incorporated in elctrostatics already, and the paper does not mention it

reisalex commented:
Hi @mengqvist I see a lot of great progress! Did you get a chance to review the following test set from the original 1994 Nat paper?
https://github.com/reisalex/pyAGADIR/blob/main/tests/test_model_NSB_1994.py
https://github.com/reisalex/pyAGADIR/pull/3#issuecomment-2351777594

I'll happily merge the PR once I see a validation on this test set! And we can push to pypi.

rcrehuet committed:
Corrected distance to N and C cap (#16)

* Corrected distance to N and C cap

* electrostatic energy should be devided by 4*pi.
Also removed some print statements.

mengqvist committed:
Make modifications to account for N- and C-terminal chemical modifications. (#17)

* add development environment

* update with installation information

* run black to format code

* run isort

* change n- and d-termenal capping residues to be arguments instead of being part of the peptide

* fix indexing for macrodipole function

* use new convention of passing the terminal modification as arguments

* update capping energies to work with get_capping_status function

* modify hydrogen bonding to account for n- and c-terminal modifications

* ensure terminal modifications are passed

* clean up and re-name function

* refactor engergy functions into a separate class

* change y axis

* initial commit

* refactor to centralize ionization calculations

* change scale back

* fix intrinsic energy bug introduced when refactoring

* consolidate print statements

* remove unnessecary peptide

* change sidechain electrostatics to a matrix, to enable per-residue assignments

* make calculate_term_dipole_interaction_energy be a method of the EnergyCalculator class

* fix indexing bug that prevented sidechain-dipole energies from being calculated

* skip uncharged residues in electrostatics calculation

* update electrostatics

* add documentation

* implement fix that counts shorter helices when there are N- or C-terminal modifications, updated get_helix function to allow shorter helices when there are modifications

* re-organize data folder to separate validation from parameters

* improve explanation of the validation

* digitize figure 4 data and plot, improve README

* fix figure size and position

* move figures inside data folder

* correct figures path

* add supplement

* remove old tables

* clean up code

* move chemistry functions to separate script, clean up in energy functions

* re-plot

* add highlight indicating that succinylation still leads to charged terminus

---------

Co-authored-by: Martin Engqvist <martin.engqvist@mpyascitech.com>

mengqvist committed:
Capping (#18)

* add development environment

* update with installation information

* run black to format code

* run isort

* change n- and d-termenal capping residues to be arguments instead of being part of the peptide

* fix indexing for macrodipole function

* use new convention of passing the terminal modification as arguments

* update capping energies to work with get_capping_status function

* modify hydrogen bonding to account for n- and c-terminal modifications

* ensure terminal modifications are passed

* clean up and re-name function

* refactor engergy functions into a separate class

* change y axis

* initial commit

* refactor to centralize ionization calculations

* change scale back

* fix intrinsic energy bug introduced when refactoring

* consolidate print statements

* remove unnessecary peptide

* change sidechain electrostatics to a matrix, to enable per-residue assignments

* make calculate_term_dipole_interaction_energy be a method of the EnergyCalculator class

* fix indexing bug that prevented sidechain-dipole energies from being calculated

* skip uncharged residues in electrostatics calculation

* update electrostatics

* add documentation

* implement fix that counts shorter helices when there are N- or C-terminal modifications, updated get_helix function to allow shorter helices when there are modifications

* re-organize data folder to separate validation from parameters

* improve explanation of the validation

* digitize figure 4 data and plot, improve README

* fix figure size and position

* move figures inside data folder

* correct figures path

* add supplement

* remove old tables

* clean up code

* move chemistry functions to separate script, clean up in energy functions

* re-plot

* add highlight indicating that succinylation still leads to charged terminus

* clarify what terminal modifications do

* account for the negative charge of succinyl

* included interaction between helix macrodipole and charged sidechain in coiled state. Fig4 looks much better, Fig3 looks worse. Intrinsic energies are too high, capping energies are too low.

* add additional checking functions

* start implementing a new class for pre-computing distances and charges

* modify class to work with current version of energies.py

* modified energy calculation functions to work with the new variables

* finished first draft of precomputing modified ionization states, termini still missing though

* simplify logic dealing with distances greater than what is in the tables

* remove i and j as arguments to energy functions

* add to explanation about the helix indexing

* add a basic set of test for the cheemistry functions

* rename debue function to reflect that it is kappa that is returned, not the inverse of it which is the screening length

* add tests for the precompute class

* simplify the finding of indexes inside and outside of the helix

* simplify code and remove the awful _are_terminal_residues_capping function

* simplify terminal macrodipole code and remove _calculate_terminal_energy method

* ensure that terminal charges get modified based on electrostatics

* finish code refactoring, everything should work now

* make new tests

* update package details

* change  whitespace

---------

Co-authored-by: Martin Engqvist <martin.engqvist@mpyascitech.com>

mengqvist committed:
Added many more validation examples (#19)

* add lacroix to figure names

* add data from huygues_despointes to the validation, re-name the lacroix figures

* add data from huygues_despointes to the validation

* add data from munoz 1997

* update figures

* update fontsizes

* fix implementation bug for how pKa is adjusted, it should now correctly account for acid and basic residues

* start implementing munoz 1995 validation

* fix schellman motif bug, the code was trying to get a table 3 value for Am

* fix so that temperature input check is on zero degrees c

* change int to float

* add temperature scaling of hbond

* finalize the temperature validation plot, they look bad

---------

Co-authored-by: Martin Engqvist <martin.engqvist@mpyascitech.com>

mengqvist committed:
apply temperature scaling to hydrophobic effects

mengqvist committed:
first implementation actually making use of table V

mengqvist committed:
fixes to electrostatic interactions, temperature-dependence of capping energies

mengqvist committed:
exclude acylation and amidation caps from helix averaging

mengqvist committed:
fix entropic correction formula

mengqvist committed:
use refactored version of _assign_modified_ionization_states, ensure distances are obtained from pre-computed table

mengqvist committed:
update tests to match the devide by 3 instead of 4, and add distance checks

mengqvist committed:
was polluting the tests with the fake distance tables, should be fixed now

mengqvist committed:
add fix to N-terminal distance, update N-terminal pKa and make tests to validate macrodipole energies

mengqvist committed:
additional documentation files

mengqvist committed:
remove terminal-sidechain interactions if the terminal is not immediately at the helix end

mengqvist committed:
update how distances are assigned

mengqvist committed:
add script to get scores indicating how well the predictions match real data

mengqvist committed:
cycleresearch changes aimed at closer matching benchmarks

mengqvist committed:
fix terminal formulas

mengqvist committed:
more automated coding fixes

mengqvist committed:
Y inonisation fix and terminal macrodipole distance cutoff

mengqvist committed:
temperature fixes

mengqvist committed:
charged side-chain interactions did not account for coiled state, fixed now
